### PR TITLE
[PROTOCOL-32] Validate Balance/Allowance Before Executing transfer/transferForm

### PR DESCRIPTION
**Pull request by sypeer**
It is:
- To verify whether the contract has enough balance or allowance and display user-friendly error messages to the final user instead, of non-readable error messages

It includes:
- "require" lending token balance validation in the tokenTransfer function in LendingPool.sol
- "require" lending token allowance validation in the tokenTransferFrom function in LendingPool.sol
- Unit test for new changes

Screen Shot 2020-07-27 at 4 27 37 PM

### CODE REVIEW

File: test/base/LendingPoolRepayTest.js [hunk 1]
```diff
@@ -41,20 +41,21 @@ contract('LendingPoolRepayTest', function (accounts) {
...
-        _7_cTokenNotSupported_notLoan: [accounts[1], accounts[2], false, true, 10, false, 'ADDRESS_ISNT_LOANS_CONTRACT', true],
-        _8_cTokenNotSupported_transferFail: [accounts[1], loansAddress, false, false, 200, false, "LENDING_TRANSFER_FROM_FAILED", true],
+        _1_cTokenSupported_basic: [accounts[1], loansAddress, true, true, 10, false, 1000, undefined, false],
+        _2_cTokenSupported_notLoan: [accounts[1], accounts[2], true, true, 10, false, 1000, 'ADDRESS_ISNT_LOANS_CONTRACT', true],
+        _3_cTokenSupported_transferFail: [accounts[1], loansAddress, true, false, 200, false, 1000, "LENDING_TRANSFER_FROM_FAILED", true],
+        _4_cTokenSupported_compoundFail: [accounts[1], loansAddress, true, true, 10, true, 1000, 'COMPOUND_DEPOSIT_ERROR', true],
+        _6_cTokenNotSupported_basic: [accounts[1], loansAddress, false, true, 10, false, 1000, undefined, false],
+        _7_cTokenNotSupported_notLoan: [accounts[1], accounts[2], false, true, 10, false, 1000, 'ADDRESS_ISNT_LOANS_CONTRACT', true],
+        _8_cTokenNotSupported_transferFail: [accounts[1], loansAddress, false, false, 200, false, 1000, "LENDING_TRANSFER_FROM_FAILED", true],
```
Comment (salazarguille) [inline test/base/LendingPoolRepayTest.js:50]: Is it possible to add a new scenario like _6_ but the allowance 0 (or less than amountToRepay) to get an allowance error?

File: test/base/LendingPoolWithdrawTest.js [hunk 2]
```diff
@@ -33,14 +35,15 @@ contract('LendingPoolWithdrawTest', function (accounts) {
     });
 
     withData({
-        _1_basic: [accounts[0], true, 10, false, undefined, false],
-        _2_transferFail: [accounts[1], false, 50, false, 'LENDING_TRANSFER_FAILED', true],
-        _3_compoundFail: [accounts[1], true, 50, true, 'COMPOUND_WITHDRAWAL_ERROR', true],
+        _1_basic: [accounts[0], true, 10, false, 1000, undefined, false],
+        _2_transferFail: [accounts[1], false, 50, false, 1000, 'LENDING_TRANSFER_FAILED', true],
+        _3_compoundFail: [accounts[1], true, 50, true, 1000, 'COMPOUND_WITHDRAWAL_ERROR', true],
```
Comment (salazarguille) [inline test/base/LendingPoolWithdrawTest.js:40]: Is it possible to add a new scenario like _1_ but the allowance 0 (or less than amountToWithdraw) to get a balance error?

**Comment by salazarguille** [comment]
Good job! I added two comments about new scenarios.

File: test/base/LendingPoolRepayTest.js (hunk 1 above)
Comment (sypeer) [reply test/base/LendingPoolRepayTest.js:50]: Added

File: test/base/LendingPoolWithdrawTest.js (hunk 2 above)
Comment (sypeer) [reply test/base/LendingPoolWithdrawTest.js:40]: Added

**Pull request merged**